homeassistant/components/zha/config_flow.py

```diff
@@ -615,34 +624,71 @@ async def async_step_zeroconf(
     ) -> ConfigFlowResult:
         """Handle zeroconf discovery."""
 
-        # Hostname is format: livingroom.local.
-        local_name = discovery_info.hostname[:-1]
-        port = discovery_info.port or DEFAULT_ZHA_ZEROCONF_PORT
+        # Transform legacy zeroconf discovery into the new format
+        if discovery_info.type != ZEROCONF_SERVICE_TYPE:
+            port = discovery_info.port or LEGACY_ZEROCONF_PORT
 
-        # Fix incorrect port for older TubesZB devices
-        if "tube" in local_name and port == ESPHOME_API_PORT:
-            port = DEFAULT_ZHA_ZEROCONF_PORT
+            # Fix incorrect port for older TubesZB devices
+            if (
+                "tube" in discovery_info.name
+                and port == LEGACY_ZEROCONF_ESPHOME_API_PORT
+            ):
+                port = LEGACY_ZEROCONF_PORT
 
-        if "radio_type" in discovery_info.properties:
-            self._radio_mgr.radio_type = self._radio_mgr.parse_radio_type(
-                discovery_info.properties["radio_type"]
+            # Determine the radio type
+            if "radio_type" in discovery_info.properties:
+                radio_type = self._radio_mgr.parse_radio_type(
+                    discovery_info.properties["radio_type"]
+                )
+            elif "efr32" in discovery_info.name:
+                radio_type = RadioType.ezsp
+            elif "zigate" in discovery_info.name:
```

Comment:
Since we're using `discovery_info.name` multiple times, I'd store the attribute in a local variable.